### PR TITLE
[Merged by Bors] - feat: Dirichlet character

### PR DESCRIPTION
**Pull request by mo271**
- [x] depends on: #7013 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
This is part of a port of 
https://github.com/laughinggas/p-adic-L-functions/blob/main/src/dirichlet_character/basic.lean

**Comment by MichaelStollBayreuth** [comment]
Is there a specific reason for not using `MulChar (ZMod n) R` as the definition of `DirichletCharacter R n`?

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+## Tags
+
+dirichlet character, multiplicative character
+-/
+-- TODO: move to Data.ZMod.Basic?!
+@[simp]
+lemma cast_hom_self {n : ℕ} : ZMod.castHom dvd_rfl (ZMod n) = RingHom.id (ZMod n) :=
+  RingHom.ext_zmod (ZMod.castHom dvd_rfl (ZMod n)) (RingHom.id (ZMod n))
+
+/-- The Dirichlet character of level `n`. -/
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:33]: Better "The type of Dirichlet characters of level `n`".
(The current version sounds like there is exactly one Dirichlet character of level `n`, which is `DirichletCharacter R n`.)

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+(x : R) : χ.toMonoidHom x = χ x := by solve_by_elim
+end MulChar
+
+open MulChar
+variable {R : Type} [CommMonoidWithZero R] {n : ℕ} (χ : DirichletCharacter R n)
+
+namespace DirichletCharacter
+lemma ofUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
+  ofUnitHom χ a = χ ha.unit := by
+  conv_lhs => rw [← (IsUnit.unit_spec ha)]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:49]: `conv_lhs => rw [← ha.unit_spec]` should also work.

**Comment by laughinggas** [comment]
> Is there a specific reason for not using `MulChar (ZMod n) R` as the definition of `DirichletCharacter R n`?

I thought it makes sense to give it a separate name since there will be a few files on it. Also, I think `MulChar` is more in sync with `asso_dirichlet_character`, which has now been replaced in terms of `ofUnitHom`.

**Comment by laughinggas** [comment]
> Also, there is maybe some ambiguity as to what should be the correct definition of a "Dirichlet character mod n". I'm more used to it being defined as a map from the integers to (some monoid with zero) R that is obtained as the composition of the reduction map Z→Z/nZ and a multiplicative character on Z/nZ. Of course, both versions are equivalent, and it is perhaps a question of what is more convenient...

I have at least 300-500 lines of code (in the folder `dirichlet_character` here : https://github.com/laughinggas/p-adic-L-functions/tree/main/src/dirichlet_character )  based on this definition. It comes with a lot of nice properties such as `change_level`, which is a monoid homomorphism. One can choose to make a second definition as a function from the integers (I am happy to contribute to that!), and show the equivalence of the two definitions.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:33]: Yes, much better, done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:49]: done

**Comment by MichaelStollBayreuth** [comment]
> > Is there a specific reason for not using `MulChar (ZMod n) R` as the definition of `DirichletCharacter R n`?
> 
> I thought it makes sense to give it a separate name since there will be a few files on it. Also, I think `MulChar` is more in sync with `asso_dirichlet_character`, which has now been replaced in terms of `ofUnitHom`.

It certainly makes sense to give it a separate name. What I mean is why not define it as
```lean
@[reducible] def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
```
instead? The rationale being that in many (most?) applications, you also want to evaluate the characters at non-units. (And, of course, that there is already an API for `MulChar`s.)

**Comment by MichaelStollBayreuth** [comment]
> > Also, there is maybe some ambiguity as to what should be the correct definition of a "Dirichlet character mod n". I'm more used to it being defined as a map from the integers to (some monoid with zero) R that is obtained as the composition of the reduction map Z→Z/nZ and a multiplicative character on Z/nZ. Of course, both versions are equivalent, and it is perhaps a question of what is more convenient...
> 
> I have at least 300-500 lines of code (in the folder `dirichlet_character` here : https://github.com/laughinggas/p-adic-L-functions/tree/main/src/dirichlet_character ) based on this definition. It comes with a lot of nice properties such as `change_level`, which is a monoid homomorphism. One can choose to make a second definition as a function from the integers (I am happy to contribute to that!), and show the equivalence of the two definitions.

After some consideration, I think it is best to use `MulChar (ZMod n) R` (as opposed to maps from integers to `R`), since if `χ` is an object of this type, `χ a` for an integer `a` just works. And, as you indicate, this definition seems to be better suited for proving properties (which, I think, comes down to the fact that the units of `ZMod n` are a thing, whereas their preimage in the integers isn't really).

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  simp only [OneHom.toFun_eq_coe, MonoidHom.toOneHom_coe] at this
+  rw [← coe_toMonoidHom, this]
+
+lemma ofUnitHom_eq_iff (ψ : DirichletCharacter R n) :
+  χ = ψ ↔ ofUnitHom χ = ofUnitHom ψ := by simp only [ofUnitHom_eq, EmbeddingLike.apply_eq_iff_eq]
+
+lemma ofUnitHom_eval_sub (x : ZMod n) :
+  ofUnitHom χ (n - x) = ofUnitHom χ (-x) := by simp
+
+lemma isPeriodic (m : ℕ) (hm : n ∣ m) (a : ℤ) :
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:64]: `m` could be implicit here (it can be deduced from `hm`).

**Comment by laughinggas** [comment]
> > > Is there a specific reason for not using `MulChar (ZMod n) R` as the definition of `DirichletCharacter R n`?
> > 
> > 
> > I thought it makes sense to give it a separate name since there will be a few files on it. Also, I think `MulChar` is more in sync with `asso_dirichlet_character`, which has now been replaced in terms of `ofUnitHom`.
> 
> It certainly makes sense to give it a separate name. What I mean is why not define it as
> 
> ```lean
> @[reducible] def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
> ```
> 
> instead? The rationale being that in many (most?) applications, you also want to evaluate the characters at non-units. (And, of course, that there is already an API for `MulChar`s.)

If you look at the following : https://github.com/laughinggas/p-adic-L-functions/blob/88c5750b71f90ae9a9964968f4eabcb994d6d279/src/dirichlet_character/basic.lean#L54 , I think the definition of `MulChar` is related more to `asso_dirichlet_character`. I removed the definition of `asso_dirichlet_character` from this port and have instead used the properties of `MulChar` instead.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 3 above)
Comment (mo271) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:64]:
```suggestion
lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) :
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+- definition of conductor
+
+## Tags
+
+dirichlet character, multiplicative character
+-/
+-- TODO: move to Data.ZMod.Basic?!
+@[simp]
+lemma cast_hom_self {n : ℕ} : ZMod.castHom dvd_rfl (ZMod n) = RingHom.id (ZMod n) :=
+  RingHom.ext_zmod (ZMod.castHom dvd_rfl (ZMod n)) (RingHom.id (ZMod n))
```
Comment (mo271) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:31]:
```suggestion
```
```suggestion
-- TODO: move to Data.ZMod.Basic?!
@[simp]
lemma cast_hom_self {n : ℕ} : ZMod.castHom dvd_rfl (ZMod n) = RingHom.id (ZMod n) :=
  RingHom.ext_zmod (ZMod.castHom dvd_rfl (ZMod n)) (RingHom.id (ZMod n))
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 3 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

**Comment by mo271** [comment]
Defining it as `(ZMod n)ˣ →* Rˣ` has the conceptual advantage of defining it on the smallest possible object, instead of either on `ZMod n` or even `\Z` with extra conditions. I agree that we should have convenient api to switch between the different incarnations.

**Comment by MichaelStollBayreuth** [comment]
OK; I guess the one reason why you want to work with homomorphisms on the unit group is that this gives you the correct notion of changing the level. I would suggest using the `MulChar` definition as the main one and then go via the equivalence to "unit homs" (`toUnitHom`/`ofUnitHom`) to do the level change etc:

````lean
def ZMod.units_map {m : ℕ} (hm : n ∣ m) : (ZMod m)ˣ →* (ZMod n)ˣ := Units.map (ZMod.castHom hm (ZMod n))

@[reducible]
def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R

namespace DirichletCharacter

noncomputable def change_level {R : Type} [CommMonoidWithZero R] {n m : ℕ} (hm : n ∣ m) :
    DirichletCharacter R n →* DirichletCharacter R m :=
  { toFun := fun ψ ↦ MulChar.ofUnitHom (ψ.toUnitHom.comp (ZMod.units_map hm)),
    map_one' := by ext; simp,
    map_mul' := fun ψ₁ ψ₂ ↦ by ext; simp }
```

**Comment by laughinggas** [comment]
> OK; I guess the one reason why you want to work with homomorphisms on the unit group is that this gives you the correct notion of changing the level. I would suggest using the `MulChar` definition as the main one and then go via the equivalence to "unit homs" (`toUnitHom`/`ofUnitHom`) to do the level change etc:
> 
> ```lean
> def ZMod.unitsMap {n m : ℕ} (hm : n ∣ m) : (ZMod m)ˣ →* (ZMod n)ˣ := Units.map (ZMod.castHom hm (ZMod n))
> 
> @[reducible]
> def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
> 
> namespace DirichletCharacter
> 
> noncomputable def change_level {R : Type} [CommMonoidWithZero R] {n m : ℕ} (hm : n ∣ m) :
>     DirichletCharacter R n →* DirichletCharacter R m :=
>   { toFun := fun ψ ↦ MulChar.ofUnitHom (ψ.toUnitHom.comp (ZMod.unitsMap hm)),
>     map_one' := by ext; simp,
>     map_mul' := fun ψ₁ ψ₂ ↦ by ext; simp }
> ```
> 
> (I was working on that while you were already answering; this refers to the earlier point. EDIT: (hopefully) corrected `units_map` to `unitsMap`...)

`change_level` is not the only reason I wish to work with the units. I am using Dirichlet characters to define the p-adic L-function. The definition of the p-adic L-function (attached below) and its properties (which is about 3000-5000 lines of code) depend on the Dirichlet character being defined on the units. I understand that this is equivalent to the definition you have given, however, it is not the same. Your definition coincides with the definition of the associated Dirichlet character, `asso_dirichlet_character` in the p-adic L-functions repository, which I reworded in terms of `MulChar.ofUnitHom`. The main reason why I am hesitant to completely remove the units is because it will involve rewriting a lot of proofs going ahead.

image

image

**Comment by laughinggas** [comment]
I can change the definition to be the following :
`def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar.toUnitHom (ZMod n) R `
but given the equivalence `equivToUnitHom`, I think it might be best to keep the original definition?

Your help is much appreciated, thank you!

**Review by kbuzzard — COMMENTED** [review]
My initial instinct is the same as Michael's. With Dirichlet characters you want to write things like $\sum \chi(n)n^{-s}$ for the L-function, and this will involve some pain if the Dirichlet character is only defined on the units (whereas there will be a coercion from the naturals to Z/mZ). I'm not sure I understand the counterarguments yet: Moritz said that using units defines chi on the smallest object possible -- but why is that a good thing? Ashvni talks about `asso_dirichlet_character` -- this seems to be the extension of the (Z/nZ)^* definition to the (Z/nZ) definition and it looks like it's some work; conversely the restriction is a triviality. I'm not sure that it will be too painful to switch to this definition; right now Ashvni's code has dirichlet characters on the units and `asso_dirichlet_character` on Z/nZ, and Michael is suggesting defining `DirichletCharacter` to be `asso_dirichlet_character` and then using this units hom trick to get what Ashvni was calling Dirichlet characters, so if Ashvni makes Michael's suggested change then all they have to do is replace their asso_dirichlet_character by DirichletCharacter and replace their Dirichlet character by the units hom map. In fact it should make the proofs simpler, because Ashvni is using `function.extend` to go one way, and this is not so easy to work with (cf their proofs of the proof fields of `asso_dirichlet_character`) whereas with the definition on all of Z/nZ the proofs should be much easier.

Dirichlet characters won't just be used for p-adic L-functions, they'll be used for classical L-functions and character sums and various other things, and in all these applications it makes like a lot easier to just be able to evaluate them on non-units. With Michael's approach we will see `\chi n` whereas with Ashvni's we'll see `asso_dirichlet_character \chi n`.

In short I am not concerned about proofs breaking with this change -- I think that they might get easier.

**Comment by MichaelStollBayreuth** [comment]
Sorry for being slow to reply (I'm trying to finish a long and involved paper...).
But Kevin has already pretty much said what I would have written. In particular, you can applpy a `MulHom` to units via the coercions from units to the monoid, and you can get the homomorphism on units via `toUnitHom` if you need it. Also (again repeating what Kevin said), in most applications I know of, you want to apply the character also to non-units, so down the line, using `MulChar` as the primary definition should be less painful.

**Comment by mo271** [comment]
Thanks for the input @MichaelStollBayreuth and @kbuzzard! I agree that a definition from `ZMod n` might make later theorems nicer. Please have a look if what we have now makes sense, so we can build on that.

**Comment by MichaelStollBayreuth** [comment]
Looks good to me so far.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+- definition of conductor
+
+## Tags
+
+dirichlet character, multiplicative character
+-/
+
+/-- The type of Dirichlet characters of level `n`. -/
+@[reducible]
+def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
```
Comment (alexjbest) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:31]:
```suggestion
abbrev DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
```

File: Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean [hunk 6]
```diff
@@ -237,6 +237,10 @@ theorem equivToUnitHom_symm_coe (f : Rˣ →* R'ˣ) (a : Rˣ) : equivToUnitHom.s
   ofUnitHom_coe f a
 #align mul_char.equiv_unit_hom_symm_coe MulChar.equivToUnitHom_symm_coe
 
+@[simp]
+lemma coe_toMonoidHom [CommMonoid R] (χ : MulChar R R')
+    (x : R) : χ.toMonoidHom x = χ x := by rfl
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean:242]: yes, it does

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+variable {R : Type} [CommMonoidWithZero R] {n : ℕ} (χ : DirichletCharacter R n)
+
+namespace DirichletCharacter
+lemma toUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
+  χ a = χ.toUnitHom ha.unit := by simp
+
+lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
+  χ = ψ ↔ toUnitHom χ = toUnitHom ψ := by simp
+
+lemma eval_sub (x : ZMod n) :
```
Comment (alexjbest) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:42]: Is this used later? it seems a bit odd to have it as a lemma if the proof is just `by simp`

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ashvni Narayanan, Moritz Firsching
+-/
+import Mathlib.NumberTheory.LegendreSymbol.MulCharacter
+import Mathlib.Data.ZMod.Algebra
+
+/-!
+# Dirichlet Characters
+
+Let `R` be a commutatvive monoid with zero. A Dirichlet character `χ` of level `n` over `R` is a
```
Comment (alexjbest) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:12]:
```suggestion
Let `R` be a commutative monoid with zero. A Dirichlet character `χ` of level `n` over `R` is a
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  χ a = χ.toUnitHom ha.unit := by simp
+
+lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
+  χ = ψ ↔ toUnitHom χ = toUnitHom ψ := by simp
+
+lemma eval_sub (x : ZMod n) :
+  χ (n - x) = χ (-x) := by simp
+
+lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) :
+  χ (a + m) = χ a := by
```
Comment (alexjbest) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:46]:
```suggestion
    χ (a + m) = χ a := by
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 7 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:42]: yes, we plan to use it later, but we can also add it later...

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+def DirichletCharacter (R : Type) [CommMonoidWithZero R] (n : ℕ) := MulChar (ZMod n) R
+
+open MulChar
+variable {R : Type} [CommMonoidWithZero R] {n : ℕ} (χ : DirichletCharacter R n)
+
+namespace DirichletCharacter
+lemma toUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
+  χ a = χ.toUnitHom ha.unit := by simp
+
+lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:39]: will change to the reversed statement...

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 7 above)
Comment (alexjbest) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:42]: Ok I would probably call it `eval_modulus_sub` or something like that

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 10 above)
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:39]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 7 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+Authors: Ashvni Narayanan, Moritz Firsching
+-/
+import Mathlib.NumberTheory.LegendreSymbol.MulCharacter
+import Mathlib.Data.ZMod.Algebra
+
+/-!
+# Dirichlet Characters
+
+Let `R` be a commutative monoid with zero. A Dirichlet character `χ` of level `n` over `R` is a
+multiplicative character from `ZMod n` to `Rˣ`. We then obtain some properties of `toUnitHom χ`.
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:13]:
```suggestion
multiplicative character from `ZMod n` to `R` sending non-units to 0. We then obtain some properties of `toUnitHom χ`, the restriction of `χ` to a group homomorphism `(ZMod n)ˣ →* Rˣ`.
```

**Review by kbuzzard — APPROVED** [review]

**Comment by riccardobrasca** [comment]
@MichaelStollBayreuth Do you think this can be merged now?

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+lemma toUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
+  χ a = χ.toUnitHom ha.unit := by simp
+
+lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
+  toUnitHom χ = toUnitHom ψ ↔ χ = ψ := by simp
+
+lemma eval_modulus_sub (x : ZMod n) :
+  χ (n - x) = χ (-x) := by simp
+
+lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:46]: Maybe we want to use [Function.Periodic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Periodic.html#Function.Periodic) (to have all the API)?

**Review by riccardobrasca — APPROVED** [review]

**Comment by MichaelStollBayreuth** [comment]
> @MichaelStollBayreuth Do you think this can be merged now?

Yes.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 12 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done: now using `Function.Periodic`

Comment (riccardobrasca) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: Looking at stuff like [Real.sin_periodic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.html#Real.sin_periodic) I don't see a clear convention.. maybe just `periodic`, but I don't think it is very important, we can always change it later.

Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: ok, changing it to `periodic`

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 13]
```diff
@@ -43,8 +44,8 @@ lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
 lemma eval_modulus_sub (x : ZMod n) :
   χ (n - x) = χ (-x) := by simp
 
-lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) :
-    χ (a + m) = χ a := by
+lemma isPeriodic {m : ℕ} (hm : n ∣ m) (a : ℤ) : Function.Periodic χ m := by
+  intro a
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:48]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import Mathlib.Algebra.Periodic
+import Mathlib.NumberTheory.LegendreSymbol.MulCharacter
+import Mathlib.Data.ZMod.Algebra
+
+/-!
+# Dirichlet Characters
+
+Let `R` be a commutative monoid with zero. A Dirichlet character `χ` of level `n` over `R` is a
+multiplicative character from `ZMod n` to `R` sending non-units to 0. We then obtain some properties
+  of `toUnitHom χ`, the restriction of `χ` to a group homomorphism `(ZMod n)ˣ →* Rˣ`.
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:15]: fixed
